### PR TITLE
Fix FBX runtime import

### LINKED ISSUE #96029: UFBX importer doesn't work in-game [Godot 4.3]
### Tested versions

4.3

### System information

manjaro linux godot 4.3-stable (this doesnt matter since this "bug" is on all platforms)

### Issue description

the new ufbx importer for godot 4.3 cant import fbx files at runtime but gltf importer can 

O. some notes:
1. the engine can import gltf at runtime but not fbx
2. the engine can import both gltf and fbx in the editor
3. at runtime loading fbx file gives an unavailable error
4. even tho now with godot 4.3 we have fbx document just like gltf document but it looks like its not working/implemented yet?

O. after some tests:
i found is that it actually (kind of) imports everything correctly but all meshes are importer mesh type with no data
so.... maybe importer mesh doesn't work in-game??

im currently looking at the source code to check whats going on with that
i have a small experience with c++ and i'd love to help (since this matters alot to me)

### Steps to reproduce

just try the MRP i gave
run the project and it wont show anything (even tho im trying to load fbx with fbxDocument just like gltfDocument)

### Minimal reproduction project (MRP)

[MRP.zip](https://github.com/user-attachments/files/16736371/MRP.zip)

### LINKED ISSUE #96043: UFBX runtime loader not working
### Tested versions

Godot v4.3.stable

### System information

Linux Mint 21 (Vanessa) - X11 - GLES3 (Compatibility) - NVIDIA GeForce RTX 3060 (nvidia; 535.183.01) - AMD Ryzen 5 2600X Six-Core Processor (12 Threads)

### Issue description

It doesn't work to load a FBX model at runetime in gdscript. Not correct, the loading works partial as you could see in the attached debugger images. But nothing is displayed in the 3D view.

### Steps to reproduce

1) unzip the MRP project
2) change "PREFIX_DIR" in Viewer.gd to an absolute path in your system where you unzipped the MRP project
3) start the project and see that only the GLTF file is loaded and the FBX files produces an error
4) look into the debugger for the FBX object and see that there're "ImportedMeshInstance3D" where in the GLTF loaded file are "MeshInstance3D"

### Minimal reproduction project (MRP)

[github_ufbx_bug.zip](https://github.com/user-attachments/files/16737913/github_ufbx_bug.zip)

Error messages:

> E 0:00:00:0934   Viewer.gd:28 @ _load_gltf_or_fbx(): No loader found for resource: /home/andreas/Games/Assets/3d/kaykit/KayKit_Adventurers_1.0_SOURCE/Characters/fbx/knight_texture.png (expected type: Texture2D)
>   <C++-Fehler>   Method/function failed. Returning: Ref<Resource>()
>   <C++-Quelle>   core/io/resource_loader.cpp:291 @ _load()
>   <Stacktrace>   Viewer.gd:28 @ _load_gltf_or_fbx()
>                  Viewer.gd:19 @ load_fbx()
>                  Viewer.gd:14 @ _ready()
> 
> W 0:00:01:0019   Viewer.gd:30 @ _load_gltf_or_fbx(): Adding 'Knight_ArmLeft' as child to 'Skeleton3D' will make owner '' inconsistent. Consider unsetting the owner beforehand.
>   <C++-Quelle>   scene/main/node.cpp:1579 @ add_child()
>   <Stacktrace>   Viewer.gd:30 @ _load_gltf_or_fbx()
>                  Viewer.gd:19 @ load_fbx()
>                  Viewer.gd:14 @ _ready()
> 
> W 0:00:01:0019   Viewer.gd:30 @ _load_gltf_or_fbx(): Adding 'Knight_ArmRight' as child to 'Skeleton3D' will make owner '' inconsistent. Consider unsetting the owner beforehand.
>   <C++-Quelle>   scene/main/node.cpp:1579 @ add_child()
>   <Stacktrace>   Viewer.gd:30 @ _load_gltf_or_fbx()
>                  Viewer.gd:19 @ load_fbx()
>                  Viewer.gd:14 @ _ready()

.......... (each bone is reported)

ufbx_remote_details

### PR DESCRIPTION
**Pull request by Naming-things-is-hard-btw**
fixes #96029
fixes #96043

ufbx can import fbx files in the editor but not at runtime

edit: 
now it fully supports loading at runtime and in editor
it fully supports document extensions
it loads textures perfectly knowing when to load with a resourceloader and when to load with a buffer
all tested animations, bones, textures, models, surfaces they all work now

### CODE REVIEW

**Comment by Naming-things-is-hard-btw** [comment]
btw im open to any improvements on this since this isn't a real solution but it works

File: modules/fbx/fbx_document.cpp [hunk 1]
```diff
@@ -2137,9 +2137,30 @@ Node *FBXDocument::generate_scene(Ref<GLTFState> p_state, float p_bake_fps, bool
...
+		}
+		Ref<Mesh> imported_mesh = cast_to<ImporterMeshInstance3D>(child)->get_mesh()->get_mesh();
+		MeshInstance3D *meshinst = memnew(MeshInstance3D);
+		child->replace_by(meshinst);
+		meshinst->set_name(child->get_name());
+		meshinst->set_transform(child->get_transform());
+		meshinst->set_mesh(imported_mesh);
+		meshinst->set_skin(child->get_skin());
+		meshinst->set_skeleton_path(child->get_skeleton_path());
+		memfree(child);
```
Comment (AThousandShips) [inline modules/fbx/fbx_document.cpp:2160]:
```suggestion
		memdelete(child);
```
You should use this method for `Object`s

Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp]: i did it...
i will do a test build to try if there are bugs
btw i found out that it still gives an error message but it imports everything perfectly so yea
also the root node has no name
i will try fo fix that too

**Comment by fire** [comment]
In gltf document we have a gltf extension system that converts this. But I was waiting for demand for this feature to restore it. 

I’ll post a gdscript version momentarily and lets figure out how to add this properly

**Comment by fire** [comment]
https://github.com/V-Sekai/TOOL_model_explorer/blob/cb9d112f051ba5426cd04b6e95228f03c30c744f/vsk_model_explorer/core/UIController.gd#L85

**Comment by Naming-things-is-hard-btw** [comment]
> In gltf document we have a gltf extension system that converts this. But I was waiting for demand for this feature to restore it.
> 
> I’ll post a gdscript version momentarily and lets figure out how to add this properly

interesting.. does that mean that we will use gltf extension for fbx or we will make a separate  extension system for fbx

**Comment by Naming-things-is-hard-btw** [comment]
> https://github.com/V-Sekai/TOOL_model_explorer/blob/cb9d112f051ba5426cd04b6e95228f03c30c744f/vsk_model_explorer/core/UIController.gd#L85
> 
> Note that this has a bug where I free empty nested importer mesh instances in error

we can add a checker that will skip if the children array size is zero

**Comment by Naming-things-is-hard-btw** [comment]
> https://github.com/V-Sekai/TOOL_model_explorer/blob/cb9d112f051ba5426cd04b6e95228f03c30c744f/vsk_model_explorer/core/UIController.gd#L85
> 
> Note that this has a bug where I free empty nested importer mesh instances in error

btw are you the one who worked on the importer? i was trying to talk to the one who implemented this for a very long time

**Comment by Naming-things-is-hard-btw** [comment]
> In gltf document we have a gltf extension system that converts this. But I was waiting for demand for this feature to restore it.
> 
> I’ll post a gdscript version momentarily and lets figure out how to add this properly

should i convert this into draft so we can work on it together??

**Comment by Naming-things-is-hard-btw** [comment]
> https://github.com/V-Sekai/TOOL_model_explorer/blob/cb9d112f051ba5426cd04b6e95228f03c30c744f/vsk_model_explorer/core/UIController.gd#L85
> 
> Note that this has a bug where I free empty nested importer mesh instances in error


i tried to build my version and i tried empty nested importer mesh instances and it didn't give an error since it checks for children size first in the for loop
```
//##########VVVVVVVVVVV
for (int i = 0; i < children.size(); i++) {
```

**Comment by fire** [comment]
My original plan was to add a FBXDocument extension system and then use FBX runtime import as the conversion of ImporterMesh3D to MeshInstance3D as a runtime plugin.

I think that would be the proper way extending the fbx importer. What do you think?

**Comment by Naming-things-is-hard-btw** [comment]
> That was the original design, but when we tried that the gltfdocument and fbxdocument didn't gain much from separation. We also had to duplicate the 10-20 auxiliary classes like `Ref<GLTFMesh>`.

oh i see then adding register_fbx_document_extension would be best right
even tho we would have both register fbx and register gltf in fbx document class that is still far cleaner than duplicating most of the code

**Comment by Naming-things-is-hard-btw** [comment]
> That was the original design, but when we tried that the gltfdocument and fbxdocument didn't gain much from separation. We also had to duplicate the 10-20 auxiliary classes like `Ref<GLTFMesh>`.


after more thinking about this i guess it would be better to just reuse gltf register funtion in fbx since fbx already uses alot of what gltf uses already (gltfmesh, gltfcamera, etc)

**Comment by Naming-things-is-hard-btw** [comment]
> That was the original design, but when we tried that the gltfdocument and fbxdocument didn't gain much from separation. We also had to duplicate the 10-20 auxiliary classes like `Ref<GLTFMesh>`.

okay i have implemented it finally... can you verify it please

**Comment by Naming-things-is-hard-btw** [comment]
> That was the original design, but when we tried that the gltfdocument and fbxdocument didn't gain much from separation. We also had to duplicate the 10-20 auxiliary classes like `Ref<GLTFMesh>`.

helo where did u disappear?
that aint fun :<

**Comment by fire** [comment]
I'm here. Let me review.

**Comment by fire** [comment]
Can you run `pre-commit run -a`? https://docs.godotengine.org/en/stable/contributing/workflow/pr_review_guidelines.html

File: modules/fbx/extensions/SCsub [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+Import("env")
+Import("env_modules")
+
+env_gltf = env_modules.Clone()
+
+# Godot source files
+
+env_gltf.add_source_files(env.modules_sources, "*.cpp")
+#env_gltf.add_source_files(env.modules_sources, "physics/*.cpp")
```
Comment (fire) [inline modules/fbx/extensions/SCsub:11]:
```suggestion
#env_gltf.add_source_files(env.modules_sources, "physics/*.cpp")
```
```suggestion
env_gltf.add_source_files(env.modules_sources, "*.cpp")
```

File: modules/fbx/fbx_document.cpp [hunk 3]
```diff
@@ -2136,6 +2136,27 @@ Node *FBXDocument::generate_scene(Ref<GLTFState> p_state, float p_bake_fps, bool
 			_import_animation(state, ap, i, p_trimming, p_remove_immutable_tracks);
 		}
 	}
+	//for (KeyValue<GLTFNodeIndex, Node *> E : state->scene_nodes) {
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:2139]: Needs revising.

**Comment by lyuma** [comment]
@Naming-things-is-hard-btw 
> helo where did u disappear?
> that aint fun :<

Please try to treat contributors with respect and refrain from posting comments which do not contribute new information (especially including unrelated quote reply in each comment). It makes it hard to follow the conversation.

Things don't always get reviewed or merged immediately, so please be patient. Reviewers have to sleep, too.

**Comment by Naming-things-is-hard-btw** [comment]
> I'm here. Let me review.
> 
> Edited:
> 
> Was away over the weekend.

**Comment by Naming-things-is-hard-btw** [comment]
> I'm here. Let me review.
> 
> Edited:
> 
> Was away over the weekend.

thanks for your time bro

btw i thought that it was complete i tried some tests and i cant import the images if the fbx model had external images
for example it tries to import png files using resource loader
is importing images also an extension or is it another thing

also i noticed that my coding style is a bit different than what godot coding style is so it refuses to accept the code how can i fix this?

**Comment by fire** [comment]
Try the earlier instructions about installing pre-commit

**Comment by fire** [comment]
> Import the images if the fbx model had external images. for example it tries to import png files using resource loader. 
>
> Is importing images also an extension or is it another thing?

We should be using the low level image importers at runtime I think, need to review the code. It's been a while since I looked at the gltf document (fbx document) image import.

**Comment by Naming-things-is-hard-btw** [comment]
> > Import the images if the fbx model had external images. for example it tries to import png files using resource loader.
> > Is importing images also an extension or is it another thing?
> 
> We should be using the low level image importers at runtime I think, need to review the code. It's been a while since I looked at the gltf document (fbx document) image import.

im searching in gltf too i think theres an extension that does this
currently i found ktx and webp importers as extensions but not png

File: modules/fbx/extensions/SCsub (hunk 2 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/extensions/SCsub]: this broke the building process reverting changes

**Comment by Naming-things-is-hard-btw** [comment]
> Can you run `pre-commit run -a`? https://docs.godotengine.org/en/stable/contributing/workflow/pr_review_guidelines.html
> 
> https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#doc-code-style-guidelines

i did it and now its formatted but it looks like i broke the .net version of godot silly me

**Comment by Naming-things-is-hard-btw** [comment]
okai i fixed the error now it imports fbx with the images without showing any errors [finally!!!!!!!!!!!!!!!] but still i want it to be perfect now it has full extension support and it can load at runtime without any errors
but the root node has no name and that bugs me also there are some warnings about inconsistent owners i want to fix those too

File: modules/fbx/fbx_document.cpp [hunk 4]
```diff
@@ -2127,6 +2136,9 @@ Node *FBXDocument::generate_scene(Ref<GLTFState> p_state, float p_bake_fps, bool
 		root = root->get_owner();
 	}
 	ERR_FAIL_NULL_V(root, nullptr);
+	if (root->get_name() == StringName()) {
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:2139]: Try is_empty().

Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:2139]: ok

Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:2139]: > Try is_empty().


okay i tried it this is stringname and not string so the function is_empty doesnt exist for this
i looked deeper on how gltf does this and i found that yea its done the same way i did it

**Comment by AThousandShips** [comment]
You should set up the [pre-commit hook](https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#pre-commit-hook) to make sure your code style is proper before you push

**Comment by Naming-things-is-hard-btw** [comment]
> Please set up the [pre-commit hook](https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#pre-commit-hook) to make sure your code style is proper before you push

yea i did that but i forgot to use that this time sorry [the older commit was working prefectly]

**Comment by AThousandShips** [comment]
You need to install the hook and it'll do its work without you manually activating it, that way you won't forget

**Comment by Naming-things-is-hard-btw** [comment]
> You need to install the hook and it'll do its work without you manually activating it, that way you won't forget

ok

**Comment by Naming-things-is-hard-btw** [comment]
> You need to install the hook and it'll do its work without you manually activating it, that way you won't forget

ok i installed it and i committed with it did that work

**Comment by AThousandShips** [comment]
It didn't because the changes you have that are using the wrong style are in earlier commits, you added commits after, you will need to apply the style fixes manually and after that things will work correctly if you have them installed

**Comment by Naming-things-is-hard-btw** [comment]
> Because I looked at the code and I can see that it is not following the style

what would you suggest me to do
is it possible to re-commit them

**Comment by AThousandShips** [comment]
You can either squash your commits and fix it, or just run `clang-format` on all the files you have added or modified

**Comment by Naming-things-is-hard-btw** [comment]
> You can either squash your commits and fix it, or just run `clang-format` on all the files you have added or modified

finally [i had to do it manually for the old files]

**Comment by Naming-things-is-hard-btw** [comment]
> You can either squash your commits and fix it, or just run `clang-format` on all the files you have added or modified

pls help
Screenshot_20240827_163712
i broke godots c#

**Comment by AThousandShips** [comment]
You need to add the class documentation of any class you bind, see [here](https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine)

**Comment by Naming-things-is-hard-btw** [comment]
> You need to add the class documentation of any class you bind, see [here](https://docs.godotengine.org/en/stable/contributing/documentation/updating_the_class_reference.html#updating-class-reference-when-working-on-the-engine)

ok i think that its kind of done now it imports fbx at editor and at runtime and it works without any errors i think its stable enough u maybe if u have time u can try a build of it

File: modules/fbx/fbx_document.cpp [hunk 5]
```diff
@@ -1076,11 +1075,17 @@ Error FBXDocument::_parse_images(Ref<FBXState> p_state, const String &p_base_pat
...
 			String base_dir = p_state->get_base_path();
-			Ref<Texture2D> texture = ResourceLoader::load(_get_texture_path(base_dir, path), "Texture2D");
-			if (texture.is_valid()) {
-				p_state->images.push_back(texture);
-				p_state->source_images.push_back(texture->get_image());
-				continue;
+			String tex_path = _get_texture_path(base_dir, path);
+			if (
+					tex_path.get_extension() != "png" &&
+					tex_path.get_extension() != "jpeg" &&
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:1081]: jpeg can also be spelled jpg.

File: modules/fbx/fbx_document.cpp [hunk 6]
```diff
@@ -2108,6 +2113,15 @@ Error FBXDocument::_parse(Ref<FBXState> p_state, String p_path, Ref<FileAccess>
 		WARN_PRINT(vformat("FBX: ignored %d further ufbx warnings", ignored_warning_count));
 	}
 
+	document_extensions.clear();
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:2116]: Are we sure it's ok to clear extensions?

File: modules/fbx/register_types.cpp [hunk 7]
```diff
@@ -84,5 +96,5 @@ void uninitialize_fbx_module(ModuleInitializationLevel p_level) {
 		return;
 	}
 	// TODO: 20240118 // fire
```
Comment (fire) [inline modules/fbx/register_types.cpp:98]:
```suggestion
```

File: doc/classes/FBXDocumentExtensionConvertImporterMesh.xml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="FBXDocumentExtensionConvertImporterMesh" inherits="GLTFDocumentExtension" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
```
Comment (fire) [inline doc/classes/FBXDocumentExtensionConvertImporterMesh.xml:5]: Can add a description.

File: modules/fbx/fbx_document.cpp (hunk 6 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:2116]: yes absolutely since this is the list of USED extensions for the current import and not the list of all extensions

File: doc/classes/FBXDocumentExtensionConvertImporterMesh.xml (hunk 8 above)
Comment (Naming-things-is-hard-btw) [reply doc/classes/FBXDocumentExtensionConvertImporterMesh.xml:5]: how?
also is it that important?
this will mostly be an internal part of the engine and no-one will touch it manually right?

File: modules/fbx/fbx_document.cpp (hunk 5 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:1081]: > jpeg can also be spelled jpg.

thx ill fix that

File: modules/fbx/register_types.cpp [hunk 9]
```diff
@@ -30,7 +30,9 @@
 
 #include "register_types.h"
 
+#include "extensions/fbx_document_extension_convert_importer_mesh.h"
 #include "fbx_document.h"
+#include "fbx_state.h"
```
Comment (AThousandShips) [inline modules/fbx/register_types.cpp:35]:
```suggestion
```
Already included

File: doc/classes/FBXDocumentExtensionConvertImporterMesh.xml (hunk 8 above)
Comment (AThousandShips) [reply doc/classes/FBXDocumentExtensionConvertImporterMesh.xml:5]: Should still have a description, please add it as you are exposing the class

Comment (Naming-things-is-hard-btw) [reply doc/classes/FBXDocumentExtensionConvertImporterMesh.xml:5]: > Should still have a description, please add it as you are exposing the class, if you are needing to expose the class, why is that needed btw?


u mean why exposing the class is needed? well no real reason other than that the GLTFDocumentExtensionConvertImporterMesh being exposed to the user for some reason im trying to make my code as "official" as i possibly can

**Comment by Naming-things-is-hard-btw** [comment]
ok i added doc review again? [sorry if i asked for reviews way too many times]

File: doc/classes/FBXDocumentExtensionConvertImporterMesh.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="FBXDocumentExtensionConvertImporterMesh" inherits="GLTFDocumentExtension" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	a Class that converts ImporterMeshes into Meshinstances for the fbx runtime loader
```
Comment (AThousandShips) [inline doc/classes/FBXDocumentExtensionConvertImporterMesh.xml:4]:
```suggestion
		A class that converts [ImporterMeshe]s into [MeshInstance3D]s for the FBX runtime loader.
```

File: modules/fbx/fbx_document.cpp [hunk 11]
```diff
@@ -1076,13 +1075,15 @@ Error FBXDocument::_parse_images(Ref<FBXState> p_state, const String &p_base_pat
 			memcpy(data.ptrw(), fbx_texture_file.content.data, fbx_texture_file.content.size);
 		} else {
 			String base_dir = p_state->get_base_path();
-			Ref<Texture2D> texture = ResourceLoader::load(_get_texture_path(base_dir, path), "Texture2D");
-			if (texture.is_valid()) {
-				p_state->images.push_back(texture);
-				p_state->source_images.push_back(texture->get_image());
-				continue;
+			if (!FileAccess::exists(_get_texture_path(base_dir, path))) {
```
Comment (AThousandShips) [inline modules/fbx/fbx_document.cpp:1078]: This checks that the file *doesn't* exist and then tries to load it, how does that work?

File: modules/fbx/fbx_document.cpp [hunk 12]
```diff
@@ -1076,13 +1075,15 @@ Error FBXDocument::_parse_images(Ref<FBXState> p_state, const String &p_base_pat
...
+				Ref<Texture2D> texture = ResourceLoader::load(_get_texture_path(base_dir, path));
+				if (texture.is_valid()) {
+					p_state->images.push_back(texture);
+					p_state->source_images.push_back(texture->get_image());
+					continue;
+				}
 			}
-			// Fallback to loading as byte array.
+			// This aint a Fallback anymore
```
Comment (AThousandShips) [inline modules/fbx/fbx_document.cpp:1086]:
```suggestion
			// This isn't a fallback anymore.
```

Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:1086]: yes this is the main loader now hehe

**Comment by Naming-things-is-hard-btw** [comment]
its now done. right?

**Comment by AThousandShips** [comment]
I can't speak for the answers to the questions I had so I'll leave it to fire to review and potentially approve, but style wise it looks good

When it's all done you'll have to squash your commits though and rebase and remove the merge commits you've added, but that's for later

File: modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+void FBXDocumentExtensionConvertImporterMesh::_copy_meta(Object *p_src_object, Object *p_dst_object) {
+	List<StringName> meta_list;
+	p_src_object->get_meta_list(&meta_list);
+	for (const StringName &meta_key : meta_list) {
+		Variant meta_value = p_src_object->get_meta(meta_key);
+		p_dst_object->set_meta(meta_key, meta_value);
+	}
+}
+
+Error FBXDocumentExtensionConvertImporterMesh::import_post(Ref<GLTFState> p_state, Node *p_root) {
```
Comment (fire) [inline modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp:46]: Can you apply the updates from https://github.com/godotengine/godot/pull/96133

Comment (fire) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp:46]: We're getting close to completion!

Comment (Naming-things-is-hard-btw) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp:46]: > Can you apply the updates from #96133

ok i will do that but it might take 2-3 days since im doing really important stuff like touching grass or watching water ya know every programmer has to do that at some point to stay sane

File: modules/fbx/fbx_document.cpp [hunk 14]
```diff
@@ -346,12 +346,11 @@ Error FBXDocument::_parse_nodes(Ref<FBXState> p_state) {
 		node.instantiate();
 
 		node->height = int(fbx_node->node_depth);
-
-		if (fbx_node->name.length > 0) {
+		if (fbx_node->is_root) {
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:349]: I don't think it's ok to always name the root node "RootNode". Previously we used what was the name in the fbx.

File: modules/fbx/fbx_document.cpp [hunk 15]
```diff
@@ -2461,10 +2490,12 @@ PackedByteArray FBXDocument::generate_buffer(Ref<GLTFState> p_state) {
 }
 
 Error FBXDocument::write_to_filesystem(Ref<GLTFState> p_state, const String &p_path) {
+	//todo: everything
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:2493]:
```suggestion
	//todo: everything
```

I don't know when this todo will ever be done, so it'll probably stay here forever. Would like to delete it.

File: modules/fbx/fbx_document.cpp [hunk 16]
```diff
@@ -2461,10 +2490,12 @@ PackedByteArray FBXDocument::generate_buffer(Ref<GLTFState> p_state) {
...
 	return ERR_UNAVAILABLE;
 }
 
 Error FBXDocument::append_from_scene(Node *p_node, Ref<GLTFState> p_state, uint32_t p_flags) {
+	//todo: everything
```
Comment (fire) [inline modules/fbx/fbx_document.cpp:2498]:
```suggestion
	//todo: everything
```

```suggestion
	//todo: everything
```

I don't know when this todo will ever be done, so it'll probably stay here forever. Would like to delete it.

File: modules/fbx/fbx_document.cpp (hunk 15 above)
Comment (fire) [reply modules/fbx/fbx_document.cpp:2493]: @bqqbarbhg doesn't have near plans for making a fbx exporter

File: modules/fbx/fbx_document.cpp (hunk 14 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:349]: sadly the fbx->rootname is almost always empty..
is it not implemented?

File: modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp (hunk 13 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp:46]: > Can you apply the updates from #96133

ok its done

File: modules/fbx/fbx_document.cpp (hunk 14 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/fbx_document.cpp:349]: > I don't think it's ok to always name the root node "RootNode". Previously we used what was the name in the fbx.

btw fbx_node->is_root slmost never returns true hmmmm...

**Comment by Naming-things-is-hard-btw** [comment]
@fire done.

**Comment by fire** [comment]
Can you squash into one commit?

**Comment by Naming-things-is-hard-btw** [comment]
> Can you squash into one commit?

umm how?

**Comment by AThousandShips** [comment]
See [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

**Comment by Naming-things-is-hard-btw** [comment]
> See [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

thx bro ur fast

**Comment by AThousandShips** [comment]
(I have it saved for these occasions! Also please don't call me bro thank you :) )

**Comment by Naming-things-is-hard-btw** [comment]
> (I have it saved for these occasions! Also please don't call me bro thank you :) )

it looks like i was too dumb to clone the repo and i was just making changes on a local repo sooooo i have no idea how to rebase something that isnt even on github
it tells me that im already in master

**Comment by AThousandShips** [comment]
You need to rebase on top of upstream, use `git rebase -i upstream/master`

**Comment by Naming-things-is-hard-btw** [comment]
sheeesh what have i done

**Comment by AThousandShips** [comment]
You pushed without `git push --force`, you need to do:
```
git reset --hard c6eb94a
git rebase -i upstream/master
git push -f
```

**Comment by fire** [comment]
Since you're struggling, do you want me to do the rebase operation for you?

**Comment by Naming-things-is-hard-btw** [comment]
> Since you're struggling, do you want me to do the rebase operation for you?

yes please do it

**Comment by fire** [comment]
1. I rebased the branch
2. Reverted the changes to import images. If the load cannot work, it should skip to the image loaders.

File: modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp (hunk 13 above)
Comment (fire) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.cpp:46]: Reapplied the changes.

**Comment by fire** [comment]
Incoming change will add `"FBXDocumentExtensionConvertImporterMesh",` to config.py so the xml documentation is within the fbx module folder.

**Comment by Naming-things-is-hard-btw** [comment]
> 1. I rebased the branch
> 
>     2. Reverted the changes to import images. If the load cannot work, it should skip to the image loaders.
> 
> 
> **Please check that this still works.**

i confirm it works

**Comment by fire** [comment]
Asking another reviewer to take a look.

File: modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef FBX_DOCUMENT_EXTENSION_CONVERT_IMPORTER_MESH_H
+#define FBX_DOCUMENT_EXTENSION_CONVERT_IMPORTER_MESH_H
+
+#include "modules/gltf/extensions/gltf_document_extension.h"
+#include "modules/gltf/gltf_defines.h"
+
+class FBXDocumentExtensionConvertImporterMesh : public GLTFDocumentExtension {
```
Comment (fire) [inline modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h:37]: Since FBXDocumentExtensionConvertImporterMesh extends GLTFDocumentExtension, what are your thoughts on using GLTFDocumentExtensionConvertImporterMesh directly?

Comment (fire) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h:37]: Not a fan of code duplication.

**Comment by Naming-things-is-hard-btw** [comment]
> 1. I rebased the branch
> 
>     2. Reverted the changes to import images. If the load cannot work, it should skip to the image loaders.
> 
> 
> **Please check that this still works.**

i confirm it works

> Asking another reviewer to take a look.

btw i was scared of squashing into one commit because i was scared of loosing my work on this
and soo after updating my local branch it looks like i actually did loose most of my work
did you merge the wrong commits?

since its back to the third or forth commit i made
the root node doesn't have a name
the importing breaks when you import an image file with the fbx file (it now gives an error even tho i fixed it before)
and its throws some warnings too

can you include the full branch since this isn't exactly all what i worked on

File: modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h (hunk 17 above)
Comment (Naming-things-is-hard-btw) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h:37]: > Not a fan of code duplication.

yea im also Not a fan of code duplication but sometimes code readability is more important

Comment (Naming-things-is-hard-btw) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h:37]: > Since FBXDocumentExtensionConvertImporterMesh extends GLTFDocumentExtension, what are your thoughts on using GLTFDocumentExtensionConvertImporterMesh directly?


sounds right but i think this might make things abit too spaghetti-like for anyone who wants to maintain it or extend it
if we want to remove this new class and reuse the gltf one we should comment that somewhere since this fbx importer is already confusing enough

since now fbx also supports extensions shouldn't we indicate that somewhere to other developers?
currently with this class most people can tell just by looking at the docs that there is both gltf and fbx improter extensions which inherit from gltf document extension
also it makes it easier to extend it later too and add more functionality to it

Comment (fire) [reply modules/fbx/extensions/fbx_document_extension_convert_importer_mesh.h:37]: Let me ask for a second opinion.

**Comment by AThousandShips** [comment]
What are you missing? The code right now is the same as it was before the squashes

**Comment by fire** [comment]
https://github.com/godotengine/godot/compare/e3550cb..c6eb94aa5ee1ae4400a54c5e71170afbc1da1986

**Comment by Naming-things-is-hard-btw** [comment]
yes that part exactly
its needed (at least to fix the annoying error in the editor)

**Comment by fire** [comment]
I'll try to find some time to put it back. You can also put it back and I'll do the squash.

The is root code looks right.

However I'm not sure about the check if exists then use the resource (texture2d). We typically load and check if is null

**Comment by Naming-things-is-hard-btw** [comment]
> I'll try to find some time to put it back. You can also put it back and I'll do the squash.
> 
> The is root code looks right.
> 
> However I'm not sure about the check if exists then use the resource (texture2d). We typically load and check if is null

no don't include it. i think we should discuss this more first

may i ask why exactly do we even try to load an image with a resource loader?
that function already ALWAYS fails
after reading the docs i found that godot can not load images with a resource loader since an image file is not a resource
BUT an image resource file is a resource and that is almost never the case with importing fbx at runtime
AND the function that's exactly below it [the "fall-back" one] already does that perfectly
so im not sure why that code exists and the only reason i didn't remove it was that i thought its important for some edge case or something.
is there's a way to SILENTLY load something? so the error is pushed into a variable rather than the console.
when exactly does this function NOT fail?

the only example i could think of would be a really rare edge case scenario where the user imports the fbx file from the editor (so all images get converted into .ctex resources) and then tries to import it at runtime (so double importing)

what do you think?

**Comment by fire** [comment]
Let me get back to you about the question.

**Comment by fire** [comment]
I am able to find effort to do a review this weekend. Sorry for the delay.

**Comment by Naming-things-is-hard-btw** [comment]
> I am able to find effort to do a review this weekend. Sorry for the delay.

no problem i understand, u can review it whenever u want

**Comment by Naming-things-is-hard-btw** [comment]
> Let me get back to you about the question.

a question...
when will this be merged?

**Comment by fire** [comment]
A reviewer in the area of FBX needs to approve and then a style / general architecture reviewer merges it in.

**Comment by fire** [comment]
I still like this feature for the Godot Engine 4.4 cycle but couldn't review the enhancements or convince someone else to look at it.

**Review by fire — COMMENTED** [review]
It's ok. Can you do a rebase to master? I will try in a test project. Do you have one?

**Comment by Naming-things-is-hard-btw** [comment]
meh...

**Comment by fire** [comment]
[96059.patch.txt](https://github.com/user-attachments/files/18258499/96059.patch.txt)

Salvaging
